File: backend/app/schemas/recommendation.py
import uuid
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


class RecommendationProfileCreate(BaseModel):
    preferred_genres: Optional[List[str]] = Field(
        default=None,
        description="List of preferred genres/topics",
    )
    preferred_authors: Optional[List[str]] = Field(
        default=None,
        description="List of preferred author names",
    )
    preferred_languages: Optional[List[str]] = Field(
        default=None,
        description="List of preferred language codes (e.g. ['en', 'fr'])",
    )
    min_rating: Optional[int] = Field(
        default=None,
        ge=1,
        le=5,
        description="Minimum star rating filter (1 to 5)",
    )
    max_rating: Optional[int] = Field(
        default=None,
        ge=1,
        le=5,
        description="Maximum star rating filter (1 to 5)",
    )


class RecommendationProfileUpdate(BaseModel):
    preferred_genres: Optional[List[str]] = Field(
        default=None,
        description="Updated list of preferred genres/topics",
    )
    preferred_authors: Optional[List[str]] = Field(
        default=None,
        description="Updated list of preferred author names",
    )
    preferred_languages: Optional[List[str]] = Field(
        default=None,
        description="Updated list of preferred language codes",
    )
    min_rating: Optional[int] = Field(
        default=None,
        ge=1,
        le=5,
        description="Updated minimum star rating filter (1 to 5)",
    )
    max_rating: Optional[int] = Field(
        default=None,
        ge=1,
        le=5,
        description="Updated maximum star rating filter (1 to 5)",
    )


class RecommendationProfileResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    preferred_genres: Optional[List[str]] = None
    preferred_authors: Optional[List[str]] = None
    preferred_languages: Optional[List[str]] = None
    min_rating: Optional[int] = None
    max_rating: Optional[int] = None
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BookRecommendationItem(BaseModel):
    book_id: uuid.UUID
    openlibrary_work_id: str
    title: str
    authors: List[str] = Field(default_factory=list)
    cover_url: Optional[str] = None
    first_publish_year: Optional[int] = None
    subjects: List[str] = Field(default_factory=list)
    score: float
    match_reasons: List[str] = Field(default_factory=list)
    average_rating: float = 0.0


class BookRecommendationResponse(BaseModel):
    recommendations: List[BookRecommendationItem] = Field(default_factory=list)
    total_count: int = 0

File: book-finder/src/types/recommendation.ts
export interface RecommendationProfile {
  id: string;
  user_id: string;
  preferred_genres?: string[] | null;
  preferred_authors?: string[] | null;
  preferred_languages?: string[] | null;
  min_rating?: number | null;
  max_rating?: number | null;
  updated_at: string;
}

export interface RecommendationProfileCreate {
  preferred_genres?: string[];
  preferred_authors?: string[];
  preferred_languages?: string[];
  min_rating?: number;
  max_rating?: number;
}

export interface RecommendationProfileUpdate {
  preferred_genres?: string[];
  preferred_authors?: string[];
  preferred_languages?: string[];
  min_rating?: number;
  max_rating?: number;
}

export interface BookRecommendationItem {
  book_id: string;
  openlibrary_work_id: string;
  title: string;
  authors: string[];
  cover_url?: string | null;
  first_publish_year?: number | null;
  subjects: string[];
  score: number;
  match_reasons: string[];
  average_rating: number;
}

export interface BookRecommendationResponse {
  recommendations: BookRecommendationItem[];
  total_count: number;
}
